### PR TITLE
[SPARK-13118][SQL] Expression encoding for optional synthetic classes

### PR DESCRIPTION
**Pull request by jodersky**
## What changes were proposed in this pull request?

Fix expression generation for optional types. 
Standard Java reflection causes issues when dealing with synthetic Scala objects (things that do not map to Java and thus contain a dollar sign in their name). This patch introduces Scala reflection in such cases.

This patch also adds a regression test for Dataset's handling of classes defined in package objects (which was the initial purpose of this PR).
## How was this patch tested?

A new test in ExpressionEncoderSuite that tests optional inner classes and a regression test for Dataset's handling of package objects.

### CODE REVIEW

**Comment by rxin** [comment]
cc @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 1]
```diff
@@ -476,7 +476,7 @@ object ScalaReflection extends ScalaReflection {
             // For non-primitives, we can just extract the object from the Option and then recurse.
             case other =>
               val className = getClassNameFromType(optType)
-              val classObj = Utils.classForName(className)
+              val classObj = getClassFromType(optType)
```
Comment (jodersky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: @rxin this was the culprit in your case

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala [hunk 2]
```diff
@@ -152,6 +152,8 @@ class ExpressionEncoderSuite extends PlanTest with AnalysisTest {
   productTest(InnerClass(1))
   encodeDecodeTest(Array(InnerClass(1)), "array of inner class")
 
+  encodeDecodeTest(Array(Option(InnerClass(1))), "array of optional inner class")
```
Comment (jodersky) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala]: this fails without the change in ScalaReflection

**Comment by jodersky** [comment]
Update: there is actually an issue, please see updates description

**Comment by rxin** [comment]
LGTM pending Jenkins

**Comment by rxin** [comment]
Looks like there is a real failure

**Comment by jodersky** [comment]
it's due to the way arrays are handled in option types (as classes). I'm working on a fix

**Comment by jodersky** [comment]
@rxin, your patch should work on top of this commit now

**Comment by rxin** [comment]
Thanks - merging in master.

**Pull request closed without merging**